crossdock/endtoend/handler_test.go

```diff
@@ -0,0 +1,117 @@
+package endtoend
+
+import (
+	"bytes"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+	"time"
+
+	"github.com/opentracing/opentracing-go"
+	"github.com/stretchr/testify/assert"
+
+	"github.com/uber/jaeger-client-go"
+	"github.com/uber/jaeger-client-go/config"
+)
+
+var (
+	testOperation = "testOperation"
+	testService   = "testService"
+
+	testConfig = config.Configuration{
+		Disabled: false,
+		Sampler: &config.SamplerConfig{
+			Type:               jaeger.SamplerTypeRemote,
+			Param:              1.0,
+			LocalAgentHostPort: "localhost:5778",
+		},
+		Reporter: &config.ReporterConfig{
+			BufferFlushInterval: time.Second,
+			LocalAgentHostPort:  "localhost:5775",
+		},
+	}
+
+	badConfig = config.Configuration{
+		Disabled: false,
+		Sampler: &config.SamplerConfig{
+			Type: "INVALID_TYPE",
+		},
+	}
+
+	testTraceRequest = traceRequest{
+		Operation: testOperation,
+		Tags: map[string]string{
+			"key": "value",
+		},
+		Count: 2,
+	}
+
+	testInvalidJSON = `bad_json`
+
+	testTraceJSONRequest = `
+		{
+			"operation": "testOperation",
+			"tags": {
+				"key": "value"
+			},
+			"count": 2
+		}
+	`
+)
+
+func newInMemoryTracer() (opentracing.Tracer, *jaeger.InMemoryReporter) {
+	inMemoryReporter := jaeger.NewInMemoryReporter()
+	tracer, _ := jaeger.NewTracer(
+		testService,
+		jaeger.NewConstSampler(true),
+		inMemoryReporter,
+		jaeger.TracerOptions.Metrics(jaeger.NewMetrics(jaeger.NullStatsReporter, nil)),
+		jaeger.TracerOptions.Logger(jaeger.NullLogger))
+	return tracer, inMemoryReporter
+}
+
+func TestInit(t *testing.T) {
+	handler := &Handler{}
+	err := handler.init(testConfig)
+	assert.NoError(t, err)
+}
+
+func TestInitBadConfig(t *testing.T) {
+	handler := &Handler{}
+	err := handler.init(badConfig)
+	assert.Error(t, err)
+}
+
+func TestTrace(t *testing.T) {
+	tracer, _ := newInMemoryTracer()
+
+	tests := []struct {
+		expectedCode int
+		json         string
+		handler      *Handler
+	}{
+		{http.StatusOK, testTraceJSONRequest, &Handler{tracer: tracer}},
+		{http.StatusBadRequest, testInvalidJSON, &Handler{}},
+		{http.StatusBadRequest, testTraceJSONRequest, &Handler{}}, // Tracer hasn't been initialized
```

Comment:
why failing to init a tracer translates to "bad request"? Shouldn't it be internal server error?